### PR TITLE
Deprecate package metadata, improve packaging documentation, refactor record package schema

### LINKED ISSUE #716: Split out the record and record package schemas
Currently the standard has the following schema files:

* release-schema.json
* release-package-schema.json
* record-package-schema.json

The record schema is embedded within the record package schema.

Since the structure of a record is an important concept and to make the distinction clearer between releases and records and their packages, I think it would be preferable if the record schema was split into a separate record-schema.json file.

The documentation could be updated too so that there are separate record schema and record package schema viewers, as with releases.

This would mean the docs and the schema files were consistently structured across both releases and records.

### LINKED ISSUE #1551: Package metadata: add a reference to publisher object
[Package metadata](https://standard.open-contracting.org/latest/en/schema/reference/?highlight=metadata#package-metadata) doesn't contain a reference to [publisher object](https://standard.open-contracting.org/latest/en/schema/reference/#publisher) so we can reference it

Some additional text to description like this:
Information to uniquely identify the publisher of this package.
See [Publisher](https://standard.open-contracting.org/latest/en/schema/reference/#publisher)

cc @jpmckinney @yolile

### LINKED ISSUE #1621: Deprecate most package metadata
While #1084 and #426 are postponed, in 1.2, we can still remove validation properties from package metadata (`required`, `uniqueItems`, etc.) and deprecate fields like `uri`, `publisher` (already at release level), `publishedDate`, `license`, `publicationPolicy`.

In other words, only `version`, `extensions` and `releases` / `records` are not deprecated.

We can also revert the `links` field from #1422.

There was also a comment in #1084 about re-organizing the content about packages: https://github.com/open-contracting/standard/issues/1084#issuecomment-1198826910

Note that when we review the documentation, we can probably also close

* #605
* #1551

### PR DESCRIPTION
**Pull request by duncandewhurst**
* Closes #1621 
* Closes #1551
* Closes #716

### CODE REVIEW

**Comment by duncandewhurst** [comment]
Regarding the test failures:

```
test_json.py::test_schema_valid[/home/runner/work/standard/standard/schema/record-package-schema.json-record-package-schema.json-data4]
  schema/record-package-schema.json has Error while resolving `[https://standard.open-contracting.org/schema/1__2__0/record-schema.json`:](https://standard.open-contracting.org/schema/1__2__0/record-schema.json%60:) HTTPError: 404 Client Error: Not Found for url: https://standard.open-contracting.org/schema/1__2__0/record-schema.json at properties/records/items
```

This will be resolved once 1__2__0 is released, I'm not sure what to do about it in the meantime. As a result, the [record package schema browser](https://standard.open-contracting.org/staging/1621-deprecate-package-metadata/en/schema/packaging/record_package/#schema-browser) does not display.

```
test_json.py::test_schema_valid[/home/runner/work/standard/standard/schema/record-schema.json-record-schema.json-data0]
  schema/record-schema.json includes "null" in "type" at /definitions/LinkedRelease/properties/url

test_json.py::test_schema_valid[/home/runner/work/standard/standard/schema/record-schema.json-record-schema.json-data0]
  schema/record-schema.json is missing "null" in "type" at /definitions/LinkedRelease/properties/tag

test_json.py::test_schema_valid[/home/runner/work/standard/standard/schema/record-schema.json-record-schema.json-data0]
  unused codelists: awardCriteria.csv, awardStatus.csv, classificationScheme.csv, contractStatus.csv, country.csv, currency.csv, documentType.csv, extendedProcurementCategory.csv, initiationType.csv, language.csv, linkRelationType.csv, mediaType.csv, method.csv, milestoneStatus.csv, milestoneType.csv, partyRole.csv, partyScale.csv, procurementCategory.csv, relatedProcess.csv, relatedProcessScheme.csv, submissionMethod.csv, tenderStatus.csv, unitClassificationScheme.csv
```

I think the tests need to be updated to add exceptions for the new record schema since these warnings were not previously reported when the `Record` definition was embedded in the record package schema:

**Comment by jpmckinney** [comment]
Hmm, `url` being required but nullable seems like an error/oversight. A linked release is not at all useful without a URL. Let's remove `null`.

Edit: I made some changes. The other errors shouldn't occur again (hard to test the 404 fix locally).

File: schema/record-package-schema.json [hunk 1]
```diff
@@ -92,7 +78,11 @@
         "string",
         "null"
       ],
-      "format": "uri"
+      "format": "uri",
+      "deprecated": {
+        "description": "This field is deprecated, because it isn't critical to distribute the license as data. However, the license can be expressed in the machine-readable description of the OCDS data in a data registry.",
```
Comment (jpmckinney) [inline schema/record-package-schema.json]:
```suggestion
        "description": "It is not critical to distribute the license URL as data. The page linking to the data can express the license.",
```

File: schema/record-package-schema.json [hunk 2]
```diff
@@ -101,13 +91,21 @@
         "string",
         "null"
       ],
-      "format": "uri"
+      "format": "uri",
+      "deprecated": {
+        "description": "This field is deprecated, because it isn't critical to distribute the publication policy as data. However, the publication policy can be expressed in the machine-readable description of the OCDS data in a data registry.",
```
Comment (jpmckinney) [inline schema/record-package-schema.json]:
```suggestion
        "description": "It is not critical to distribute the publication policy URL as data. The page linking to the data can express the publication policy.",
```

File: schema/record-package-schema.json [hunk 3]
```diff
@@ -81,9 +66,10 @@
           "format": "uri"
         }
       },
-      "required": [
-        "name"
-      ]
+      "deprecated": {
+        "description": "This field is deprecated in favour of the `publisher` field at release level.",
```
Comment (jpmckinney) [inline schema/record-package-schema.json]:
```suggestion
        "description": "This field is deprecated in favor of the `publisher` field in the release schema.",
```

File: schema/record-package-schema.json [hunk 4]
```diff
@@ -5,30 +5,15 @@
...
-      }
-    },
     "uri": {
       "title": "Package identifier",
       "description": "The URI of this package that identifies it uniquely in the world.",
       "type": "string",
-      "format": "uri"
+      "format": "uri",
+      "deprecated": {
+        "description": "This field is deprecated for consistency with a planned future JSONL bulk data format.",
```
Comment (jpmckinney) [inline schema/record-package-schema.json]:
```suggestion
        "description": "This field mirrored the `uri` field in the release package schema, which is deprecated.",
```

File: schema/release-package-schema.json [hunk 5]
```diff
@@ -5,37 +5,19 @@
...
-      }
-    },
     "uri": {
       "title": "Package identifier",
       "description": "The URI of this package that identifies it uniquely in the world. Recommended practice is to use a dereferenceable URI, where a persistent copy of this package is available.",
       "type": "string",
-      "format": "uri"
+      "format": "uri",
+      "deprecated": {
+        "description": "This field is deprecated for consistency with a planned future JSONL bulk data format.",
```
Comment (jpmckinney) [inline schema/release-package-schema.json]:
```suggestion
        "description": "This field was used by the `packages` field in the record package schema, which is deprecated.",
```

File: schema/record-package-schema.json [hunk 6]
```diff
@@ -101,13 +91,21 @@
...
+      }
     },
     "publishedDate": {
       "title": "Published date",
       "description": "The date that this package was published. If this package is generated 'on demand', this date should reflect the date of the last change to the underlying contents of the package.",
       "type": "string",
-      "format": "date-time"
+      "format": "date-time",
+      "deprecated": {
+        "description": "This field is deprecated for consistency with a planned future JSONL bulk data format.",
```
Comment (jpmckinney) [inline schema/record-package-schema.json]:
```suggestion
        "description": "This field repeats the maximum `date` among the releases it contains.",
```

File: docs/schema/packaging/record_package.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+````
+
+For this version of OCDS, the canonical URL for the record package schema is <https://standard.open-contracting.org/schema/1__1__5/record-package-schema.json> and for the versioned release schema is <https://standard.open-contracting.org/schema/1__1__5/versioned-release-validation-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
+
+In addition to publishing the structured metadata described by the record package schema, you ought to license your data and provide guidance for data users. For more information, see the [licensing guidance](../../guidance/publish.md#license-your-data) and [publication policy guidance](../../guidance/publish.md#finalize-your-publication-policy).
+
+```{admonition} Browsing the schema
+:class: note
+
+This page presents the record package schema in an interactive browser and in a table, with additional information in paragraphs. You can also download the canonical version of the record package schema as [JSON Schema](../../../build/current_lang/record-package-schema.json).
```
Comment (jpmckinney) [inline docs/schema/packaging/record_package.md]: We don't have additional paragraphs here.

```suggestion
This page presents the record package schema in an interactive browser and in a table. You can also download the canonical version of the record package schema as [JSON Schema](../../../build/current_lang/record-package-schema.json).
```

File: docs/schema/packaging/record_package.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+```{jsoninclude} ../../examples/merging/updates/versioned.json
+:jsonpointer:
+:title: Record package
+```
+````
+
+For this version of OCDS, the canonical URL for the record package schema is <https://standard.open-contracting.org/schema/1__1__5/record-package-schema.json> and for the versioned release schema is <https://standard.open-contracting.org/schema/1__1__5/versioned-release-validation-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
+
+In addition to publishing the structured metadata described by the record package schema, you ought to license your data and provide guidance for data users. For more information, see the [licensing guidance](../../guidance/publish.md#license-your-data) and [publication policy guidance](../../guidance/publish.md#finalize-your-publication-policy).
```
Comment (jpmckinney) [inline docs/schema/packaging/record_package.md]:
```suggestion
In addition to publishing the metadata described by the record package schema, you ought to license your data and guide data users. For more information, see the [licensing](../../guidance/publish.md#license-your-data) and [publication policy](../../guidance/publish.md#finalize-your-publication-policy) guidance.
```

File: docs/schema/packaging/index.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# Packaging
+
+In this section you will find the schema for [release packages](release_package.md) and [record packages](record_package.md), which act as envelopes for releases and records, respectively.
```
Comment (jpmckinney) [inline docs/schema/packaging/index.md]:
```suggestion
This section describes the schema for [release packages](release_package) and [record packages](record_package), which act as envelopes for releases and records, respectively.
```

File: docs/schema/packaging/record_package.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# Record Package Schema
+
+The record package schema describes the structure of the container for publishing records. The package contains important metadata.
```
Comment (jpmckinney) [inline docs/schema/packaging/record_package.md]:
```suggestion
The record package schema describes a container for records.
```

File: docs/schema/packaging/record_package.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+````{admonition} Example
+:class: hint
+
+```{jsoninclude} ../../examples/merging/updates/versioned.json
```
Comment (jpmckinney) [inline docs/schema/packaging/record_package.md:8]: The two examples are overwhelmingly not package metadata. I would prefer an example that just puts an empty object where other schemas are involved.

The examples were added to the other pages in #1482 to close #1348. They should be kept there, as these specific examples where intended to demonstrate the record and release (in the context of their package).

File: docs/schema/packaging/record_package.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+:jsonpointer:
+:title: Record package
+```
+````
+
+For this version of OCDS, the canonical URL for the record package schema is <https://standard.open-contracting.org/schema/1__1__5/record-package-schema.json> and for the versioned release schema is <https://standard.open-contracting.org/schema/1__1__5/versioned-release-validation-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
```
Comment (jpmckinney) [inline docs/schema/packaging/record_package.md]: Versioned release schema is moved to record schema.

```suggestion
For this version of OCDS, the canonical URL for the record package schema is <https://standard.open-contracting.org/schema/1__1__5/record-package-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
```

File: docs/schema/reference.md [hunk 13]
```diff
@@ -64,43 +66,20 @@ A contract for ‘Software consultancy services’ is published in a release wit
 
 ## Release structure
 
-The majority of OCDS data is held within a release structure. One or more releases can be published within a release package. Releases are made up of a number of sections, arranged in the following structure.
+The majority of OCDS data is held within a release structure. One or more releases can be published within a [release package](packaging/release_package.md). Releases are made up of a number of sections, arranged in the following structure.
```
Comment (jpmckinney) [inline docs/schema/reference.md]: Omit the `.md` everywhere, for internal links. It's only needed for anchor links `page.md#anchor`.

```suggestion
The majority of OCDS data is held within a release structure. One or more releases can be published within a [release package](packaging/release_package). Releases are made up of a number of sections, arranged in the following structure.
```

File: docs/schema/record.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Record Schema
+
+The record schema is used to check the structure of the `records` array of [record packages](packaging/record_package).
+
+For this version of OCDS, the canonical URL of the release schema is: <https://standard.open-contracting.org/schema/1__2__0/record-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
```
Comment (jpmckinney) [inline docs/schema/record.md]:
```suggestion
For this version of OCDS, the canonical URL of the record schema is: <https://standard.open-contracting.org/schema/1__2__0/record-schema.json>. Using the canonical URL guarantees that your software, documentation or other resources will always refer to the specific version of the schema with which they were authored and tested.
```

File: docs/schema/record.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+Click on schema elements to expand the tree, or use the '+' icon to expand all elements. Use { } to view the underlying schema for any section. Required fields are indicated in **bold**. [Deprecated fields](../governance/deprecation) and [multilingual fields](reference.md#language) are omitted.
+
+```{admonition} Browsing the schema
+:class: note
+
+This page presents the release schema in an interactive browser. You can also download the canonical version of the release schema as [JSON Schema](../../build/current_lang/record-schema.json) or view it as [tables](records_reference).
```
Comment (jpmckinney) [inline docs/schema/record.md]:
```suggestion
This page presents the record schema in an interactive browser. You can also download the canonical version of the record schema as [JSON Schema](../../build/current_lang/record-schema.json) or view it as [tables](records_reference).
```

**Review by jpmckinney — COMMENTED** [review]
Almost there! Just a few fixes.

Need to delete the two changelog entries for #1422. Also missing changelog entries for:
* Deprecations (the deprecated fields and any sub-fields are no longer required)
* records and releases are no longer required to be unique

Governance-wise, we can't justify the present with the future, so I changed the deprecation notices.

Please make identical or corresponding changes to the release package schema and page.

File: docs/schema/packaging/record_package.md (hunk 11 above)
Comment (duncandewhurst) [reply docs/schema/packaging/record_package.md:8]: > The two examples are overwhelmingly not package metadata. I would prefer an example that just puts an empty object where other schemas are involved.

There's no `expand` option specified so the rendered result looks the same and has the advantage of actually being valid data (unlike an empty record object):

image

Happy to replace it with a minimal example with an empty object, if you still prefer that option.

> The examples were added to the other pages in #1482 to close #1348. They should be kept there, as these specific examples where intended to demonstrate the record and release (in the context of their package).

The sections in which the examples appeared no longer exist. The examples were introduced with a sentence like "Records must be published within a [record package](https://standard.open-contracting.org/staging/1.2-dev/en/schema/record_package/)." Now that packages are optional, I figured it makes sense to provide examples demonstrating the relationship between releases/records and packages on the package reference pages. Where would you want to put the package examples on the release and record reference pages and what framing should we provide?

If we choose to leave them off the release and record pages, I think we should add an example of a record to the record reference page in order to demonstrate what a full record looks like. (The release structure section of the release reference already includes an example of a release.)

**Comment by duncandewhurst** [comment]
> Hmm, `url` being required but nullable seems like an error/oversight. A linked release is not at all useful without a URL. Let's remove `null`.

Done in https://github.com/open-contracting/standard/pull/1640/commits/80d245b62755f1dc472c0aa1f6ce4c41d66c5c51

> Edit: I made some changes. The other errors shouldn't occur again (hard to test the 404 fix locally).

The 404 error still occurs

> Need to delete the two changelog entries for #1422. 

Done in https://github.com/open-contracting/standard/pull/1640/commits/710ab21ca9c5bfc0a118e30b34450affee53124f

> Also missing changelog entries for:
> 
>     * Deprecations (the deprecated fields and any sub-fields are no longer required)
> 
>     * records and releases are no longer required to be unique

Done in https://github.com/open-contracting/standard/pull/1640/commits/d4e689743b54252cd124f86ffa9767686c9e83fb 

> Governance-wise, we can't justify the present with the future, so I changed the deprecation notices.
> 
> Please make identical or corresponding changes to the release package schema and page.

Done in https://github.com/open-contracting/standard/pull/1640/commits/3d1b39d75edd3b6aeb8bcf25d69a47021144ea32 and https://github.com/open-contracting/standard/pull/1640/commits/5b3358e8c8cd2572b7d043239fcb74636cab9398

File: docs/schema/packaging/record_package.md (hunk 11 above)
Comment (jpmckinney) [reply docs/schema/packaging/record_package.md:8]: > There's no expand option specified so the rendered result looks the same

Ah, perfect. That works.

> If we choose to leave them off the release and record pages, I think we should add an example of a record to the record reference page in order to demonstrate what a full record looks like.

Yes, sounds good.

> The sections in which the examples appeared no longer exist.

Oh, right. Your changes sound fine then.

Just a note to myself to take a look at the rendered pages when doing a full review.

**Comment by jpmckinney** [comment]
> The 404 error still occurs

Ah, `test_json.py` was trying to use `GITHUB_BASE_REF` to determine whether the PR was for an unreleased tag or not (e.g. if the base ref contained "1.2", like `1.2-dev`), but that environment variable is only available if the workflow is triggered by the `pull_request` event. The workflow has an `if` condition in order to not run twice for both the `push` and `pull_request` events (it would always run for `push`, and only for `pull_request` if the PR was on a fork).

I've removed that `if` condition. That said, I expect the PR will still fail on the `push` event, but it should pass on the `pull_request` event. Edit: I now skip the `push` event on PRs.

File: docs/schema/packaging/record_package.md (hunk 11 above)
Comment (duncandewhurst) [reply docs/schema/packaging/record_package.md:8]: > > If we choose to leave them off the release and record pages, I think we should add an example of a record to the record reference page in order to demonstrate what a full record looks like.
> 
> Yes, sounds good.

Done in https://github.com/open-contracting/standard/pull/1640/commits/ddfdc0c12406222f14870afe0ebba865b8db02d8

File: docs/schema/packaging/release_package.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
+# Release Package Reference
```
Comment (jpmckinney) [inline docs/schema/packaging/release_package.md]: To match "Record Package Schema"

```suggestion
# Release Package Schema
```

**Review by jpmckinney — APPROVED** [review]
@duncandewhurst I added a few more commits to cut other content about the deprecated fields. If you agree, you can merge.

**Comment by duncandewhurst** [comment]
The new commits look good!

**Pull request merged**